File: code/02-AsyncConsole/AsyncConsole.cs
using System;
using System.Threading.Tasks;

/// <summary>
/// Very simple demonstration of the async task. 
/// </summary>
/// <remarks>
/// Slightly modified version of http://www.dotnetperls.com/async
/// </remarks>
internal class AsyncConsole
{
  private static void Main()
  {
    while (true)
    {
      // Handle user input
      string result = Console.ReadLine();
      Console.WriteLine("You typed: {0}", result);
      // Start computation.
      Example(result);
    }
  }

  private static async void Example(string origin)
  {
    // This method runs asynchronously.
    int t = await Task.Run(() => Allocate());
    Console.WriteLine("Compute {0}: {1}", origin, t);
  }

  private static int Allocate()
  {
    // Compute total count of digits in strings
    int size = 0;
    for (int z = 0; z < 100; z++)
    {
      for (int i = 0; i < 1000000; i++)
      {
        string value = i.ToString();
        size += value.Length;
      }
    }
    return size;
  }
}

File: code/FSharp/02-AsyncConsole/AsyncConsole.fs
// Async computation example
// from http://www.dotnetperls.com/async
// implemented in F#

let allocate () = 
    let mutable size = 0
    for z in { 0..99 } do
        for i in { 0..999999 } do
            let s = string i
            size <- size + s.Length
    size

let example str = async {
        let! t = async {
            return allocate ()
        }
        printf "Compute %s: %d\n" str t
    }

let rec run start str =
    if start
      then (example str) |> Async.Start
    let r = System.Console.ReadLine()
    printf "You typed: %s\n" r
    run true r

[<EntryPoint>]
let main argv = 
    run false ""